src/onnxruntime.cc

```diff
@@ -724,43 +755,122 @@ ModelState::LoadModel(
+          cuda_options_map["gpu_mem_limit"] = string_value;
+        } else {
+          cuda_options_map["gpu_mem_limit"] =
+              std::to_string(std::numeric_limits<size_t>::max());
+        }
+
+        if (params.Find("arena_extend_strategy", &json_value)) {
+          int arena_extend_strategy = 0;
+          RETURN_IF_ERROR(TryParseModelStringParameter(
+              params, "arena_extend_strategy", &arena_extend_strategy, 0));
+          std::string string_value;
+          switch (arena_extend_strategy) {
+            case 0:
+              string_value = "kNextPowerOfTwo";
+              break;
+            case 1:
+              string_value = "kSameAsRequested";
+              break;
+            default:
+              return TRITONSERVER_ErrorNew(
+                  TRITONSERVER_ERROR_INVALID_ARG,
+                  (std::string("unsupported arena_extend_strategy value '") +
+                   std::to_string(arena_extend_strategy) + "' is requested")
+                      .c_str());
+          }
+          cuda_options_map["arena_extend_strategy"] = string_value;
+        } else {
+          cuda_options_map["arena_extend_strategy"] = "kNextPowerOfTwo";
+        }
+
+        if (params.Find("do_copy_in_default_stream", &json_value)) {
+          std::string string_value;
+          RETURN_IF_ERROR(
+              json_value.MemberAsString("string_value", &string_value));
+          cuda_options_map["do_copy_in_default_stream"] = string_value;
+        } else {
+          cuda_options_map["do_copy_in_default_stream"] = "1";
+        }
       }
     }
 
-    RETURN_IF_ORT_ERROR(ort_api->SessionOptionsAppendExecutionProvider_CUDA(
-        soptions, &cuda_options));
+    std::vector<const char*> option_names, option_values;
+    for (const auto& [key, value] : cuda_options_map) {
+      option_names.push_back(key.c_str());
+      option_values.push_back(value.c_str());
+    }
+
+    RETURN_IF_ORT_ERROR(ort_api->UpdateCUDAProviderOptions(
+        rel_cuda_options.get(), option_names.data(), option_values.data(),
+        option_values.size()));
+
+    if (stream != nullptr) {
+      RETURN_IF_ORT_ERROR(ort_api->UpdateCUDAProviderOptionsWithValue(
+          rel_cuda_options.get(), "user_compute_stream", stream));
+    }
+    RETURN_IF_ORT_ERROR(ort_api->SessionOptionsAppendExecutionProvider_CUDA_V2(
+        soptions, cuda_options));
+
+    OrtAllocator* allocator;
+    char* options;
+    RETURN_IF_ORT_ERROR(ort_api->GetAllocatorWithDefaultOptions(&allocator));
+    RETURN_IF_ORT_ERROR(ort_api->GetCUDAProviderOptionsAsString(
+        rel_cuda_options.get(), allocator, &options));
     LOG_MESSAGE(
         TRITONSERVER_LOG_VERBOSE,
         (std::string("CUDA Execution Accelerator is set for '") + Name() +
-         "' on device " + std::to_string(instance_group_device_id))
+         "' on device " + std::to_string(instance_group_device_id) +
+         std::string(" with options: ") + std::string(options))
```

Comment:
Open to discuss if we need to print out the cuda ep options. Without this I think we cannot fully test it if the options are set correctly, but might be a bit overhead for the logging.

Reply:
This looks good to me. It is verbose logging anyways.